x/lightclient/keeper/canonical_client.go

```diff
@@ -78,56 +87,67 @@ func (k Keeper) validClient(ctx sdk.Context, clientID string, cs exported.Client
 	}
 
 	expClient := k.expectedClient(ctx)
-
 	if err := types.IsCanonicalClientParamsValid(tmClientState, &expClient); err != nil {
 		return errorsmod.Wrap(err, "params")
 	}
 
-	// FIXME: No need to get all consensus states. should iterate over the consensus states
-	res, err := k.ibcClientKeeper.ConsensusStateHeights(ctx, &ibcclienttypes.QueryConsensusStateHeightsRequest{
-		ClientId:   clientID,
-		Pagination: &query.PageRequest{Limit: maxHeight},
-	})
-	if err != nil {
-		return errorsmod.Wrap(err, "cons state heights")
-	}
 	atLeastOneMatch := false
-	for _, consensusHeight := range res.ConsensusStateHeights {
-		h := consensusHeight.GetRevisionHeight()
-		if maxHeight < h {
-			break
+	csStore := k.ibcClientKeeper.ClientStore(ctx, clientID)
+	var err error
+	IterateConsensusStateDescending(csStore, func(h exported.Height) bool {
+		// skip future heights
+		if h.GetRevisionHeight() >= maxHeight {
+			return false
 		}
-		consensusState, _ := k.ibcClientKeeper.GetClientConsensusState(ctx, clientID, consensusHeight)
-		tmConsensusState, _ := consensusState.(*ibctm.ConsensusState)
-		stateInfoH, err := k.rollappKeeper.FindStateInfoByHeight(ctx, rollappId, h)
-		if err != nil {
-			return errorsmod.Wrapf(err, "find state info by height h: %d", h)
-		}
-		stateInfoHplus1, err := k.rollappKeeper.FindStateInfoByHeight(ctx, rollappId, h+1)
-		if err != nil {
-			return errorsmod.Wrapf(err, "find state info by height h+1: %d", h+1)
+
+		// iterate until we pass the fraud height
+		if h.GetRevisionHeight() < minHeight {
+			return true // break
 		}
-		bd, _ := stateInfoH.GetBlockDescriptor(h)
 
-		nextSeq, err := k.SeqK.RealSequencer(ctx, stateInfoHplus1.Sequencer)
-		if err != nil {
-			return errorsmod.Wrap(err, "get sequencer")
+		consensusState, ok := k.ibcClientKeeper.GetClientConsensusState(ctx, clientID, h)
+		if !ok {
+			return false
 		}
-		rollappState := types.RollappState{
-			BlockDescriptor:    bd,
-			NextBlockSequencer: nextSeq,
+		tmConsensusState, ok := consensusState.(*ibctm.ConsensusState)
+		if !ok {
+			return false
 		}
-		err = types.CheckCompatibility(*tmConsensusState, rollappState)
+
+		err = k.ValidateUpdatePessimistically(ctx, sInfo, tmConsensusState, h.GetRevisionHeight())
 		if err != nil {
-			return errorsmod.Wrapf(err, "check compatibility: height: %d", h)
+			err = errorsmod.Wrapf(err, "validate pessimistic h: %d", h.GetRevisionHeight())
+			return true // break
 		}
+
 		atLeastOneMatch = true
-	}
+		return true // break
+	})
 	// Need to be sure that at least one consensus state agrees with a state update
 	// (There are also no disagreeing consensus states. There may be some consensus states
 	// for future state updates, which will incur a fraud if they disagree.)
```

Comment:
This comment is no longer valid since you dont check against states less than the state info, and they wont be frauded